app/importers/unpaywall_publication_importer.rb

```diff
@@ -0,0 +1,91 @@
+# frozen_string_literal: true
+
+class UnpaywallPublicationImporter
+  def import_all
+    unless Rails.env.test?
+      pbar = ProgressBar.create(title: 'Importing publication data from Unpaywall',
+                                total: all_pubs.count)
+    end
+
+    all_pubs.find_each do |p|
+      query_unpaywall_for(p)
+      pbar.increment unless Rails.env.test?
+    end
+    pbar.finish unless Rails.env.test?
+  end
+
+  def import_new
+    unless Rails.env.test?
+      pbar = ProgressBar.create(title: 'Importing publication data from Unpaywall',
+                                total: new_pubs.count)
+    end
+
+    new_pubs.find_each do |p|
+      query_unpaywall_for(p)
+      pbar.increment unless Rails.env.test?
+    end
+    pbar.finish unless Rails.env.test?
+  end
+
+  private
+
+    def all_pubs
+      Publication.where.not(doi: nil).where.not(doi: '')
+    end
+
+    def new_pubs
+      all_pubs.where(unpaywall_last_checked_at: nil)
+    end
+
+    def get_pub(url)
+      attempts = 0
+      HTTParty.get(url).to_s
+    rescue Net::ReadTimeout, Net::OpenTimeout
+      if attempts <= 10
+        attempts += 1
+        retry
+      else
+        raise
+      end
+    end
+
+    def query_unpaywall_for(publication)
+      unpaywall_json = nil
+      find_url = URI::DEFAULT_PARSER.escape("https://api.unpaywall.org/v2/#{publication.doi_url_path}?email=openaccess@psu.edu")
+      unpaywall_json = JSON.parse(get_pub(find_url))
+
+      best_oa_location = unpaywall_json.dig('best_oa_location', 'url')
+      existing_oa_location = publication.open_access_locations.find_by(source: 'Unpaywall')
+      oa_status = unpaywall_json['oa_status']
+
+      if best_oa_location
+        if existing_oa_location
+          existing_oa_location.update!(url: best_oa_location)
+        else
+          publication.open_access_locations.create!(source: 'Unpaywall', url: best_oa_location)
+        end
+      else
+        existing_oa_location.try(:destroy)
+      end
+
+      publication.open_access_status = oa_status
+      publication.unpaywall_last_checked_at = Time.current
+      publication.save!
+
+      # Unpaywall asks that users limit requests to no more than 100,000 per day.
+      # Limiting to 1 request per second caps us at 86,400 requests per day.
+      unless Rails.env.test?
+        sleep 1
```

Comment:
Good one-liner here: `sleep 1 unless Rails.env.test?`. 🥁

Reply:
👍 I still sometimes forget about this Ruby way of writing conditionals. 😁